sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeLocalShuffleReader.scala

```diff
@@ -0,0 +1,142 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, UnknownPartitioning}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{LeafExecNode, SparkPlan}
+import org.apache.spark.sql.execution.exchange.{EnsureRequirements, ShuffleExchangeExec}
+import org.apache.spark.sql.execution.joins.{BroadcastHashJoinExec, BuildLeft, BuildRight}
+import org.apache.spark.sql.internal.SQLConf
+
+case class OptimizeLocalShuffleReader(conf: SQLConf) extends Rule[SparkPlan] {
+
+  def canUseOrRevertLocalShuffleReaderLeft(join: BroadcastHashJoinExec): Boolean = {
+    (join.buildSide == BuildRight &&  ShuffleQueryStageExec.isShuffleQueryStageExec(join.left)) ||
+      (join.buildSide == BuildRight && join.left.isInstanceOf[LocalShuffleReaderExec])
+  }
+
+  def canUseOrRevertLocalShuffleReaderRight(join: BroadcastHashJoinExec): Boolean = {
+    (join.buildSide == BuildLeft &&  ShuffleQueryStageExec.isShuffleQueryStageExec(join.right)) ||
+      (join.buildSide == BuildLeft && join.right.isInstanceOf[LocalShuffleReaderExec])
+  }
+
+  def revertLocalShuffleReader(plan: SparkPlan): SparkPlan = {
+    plan.transformDown {
+      case join: BroadcastHashJoinExec if canUseOrRevertLocalShuffleReaderRight(join) =>
+        join.copy(right = join.right.asInstanceOf[LocalShuffleReaderExec].child)
+      case join: BroadcastHashJoinExec if canUseOrRevertLocalShuffleReaderLeft(join) =>
+        join.copy(left = join.left.asInstanceOf[LocalShuffleReaderExec].child)
+    }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.optimizedLocalShuffleReaderEnabled) {
+      return plan
+    }
+
+    val optimizedPlan = plan.transformDown {
+      case join: BroadcastHashJoinExec if canUseOrRevertLocalShuffleReaderRight(join) =>
+        val localReader = LocalShuffleReaderExec(join.right.asInstanceOf[QueryStageExec])
+        join.copy(right = localReader)
+      case join: BroadcastHashJoinExec if canUseOrRevertLocalShuffleReaderLeft(join) =>
+        val localReader = LocalShuffleReaderExec(join.left.asInstanceOf[QueryStageExec])
+        join.copy(left = localReader)
+    }
+
+    val afterEnsureRequirements = EnsureRequirements(conf).apply(optimizedPlan)
+
+    val numExchanges = afterEnsureRequirements.collect {
+      case e: ShuffleExchangeExec => e
+    }.length
+
+    if (numExchanges > 0) {
```

Comment:
It's very likely that we have exchanges in the query plan, what we need to guarantee here is we don't introduce extra exchanges. We should do:
```
def numExchanges...
val numExchangesBefore = numExchanges(EnsureRequirements(conf).apply(plan))
val numExchangesAfter = numExchanges(EnsureRequirements(conf).apply(optimizedPlan))
if (numExchangesAfter > numExchangesBefore) {
  plan
} else {
  optimizedPlan
}
```